tensorflow/compiler/xla/service/gpu/convolution_thunk.cc

```diff
@@ -295,26 +297,32 @@ tensorflow::Status ConvolutionThunk::ConvolveWithTune(
 
     se::dnn::ProfileResult best_result;
     se::dnn::ProfileResult best_result_without_scratch;
-    for (se::dnn::AlgorithmType algorithm : GetAlgorithms(stream->parent())) {
-      ConvolveScratchAllocator scratch_allocator(
-          buffer_allocations.device_ordinal(),
-          buffer_allocations.memory_allocator());
-      se::dnn::ProfileResult profile_result;
-      bool launch_ok =
-          Convolve(input_descriptor, input_data, filter_descriptor, filter_data,
-                   output_descriptor, output_data, convolution_descriptor,
-                   se::dnn::AlgorithmConfig(algorithm, algorithm), stream,
-                   &scratch_allocator, &profile_result)
-              .ok();
-      if (launch_ok && profile_result.is_valid()) {
-        if (profile_result.elapsed_time_in_ms() <
-            best_result.elapsed_time_in_ms()) {
-          best_result = profile_result;
-        }
-        if (scratch_allocator.TotalAllocatedBytes() == 0 &&
-            profile_result.elapsed_time_in_ms() <
-                best_result_without_scratch.elapsed_time_in_ms()) {
-          best_result_without_scratch = profile_result;
+    std::vector<se::dnn::AlgorithmType::Index> algorithms
+        = GetAlgorithms(stream->parent());
+    for (int use_tensor_ops = 0; use_tensor_ops < 2; ++use_tensor_ops) {
```

Comment:
Much prefer to write this loop as

  for (bool use_tensor_ops : {false, true}) { ... }

But even better would be for GetAlgorithms to return a vector<se::dnn::AlgorithmType> -- then you wouldn't need to do this at all.

Reply:
Improved loop syntax.

Reply:
Did you look into making GetAlgorithms return a `vector<se::dnn::AlgorithmDesc>`?  I think that would greatly simplify things...

Reply:
This would require fixing the forward, backward filter and backward data GetAlgorithm variants. There is a fair bit of cleanup needed around this section of code in general. I think it would be better to tackle it as a separate issue.

Reply:
Added TODO to track this.